File: day10/knot_hash.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.24);
use experimental qw(signatures);

# test data
#my @list = 0..4;
#my @LENS = (3, 4, 1, 5);

# problem data
my @list = 0..255;
my @LENS = (147,37,249,1,31,2,226,0,161,71,254,243,183,255,30,70);

my $p = 0;
my $skip = 0;

for my $len (@LENS) {
    my $end = $p + $len - 1;
    if ($end <= $#list) {
        @list[$p..$end] = reverse @list[$p..$end];
    } else {
        my @tmp = (@list, @list);
        @tmp[$p..$end] = reverse @tmp[$p..$end];
        @tmp[0..($end-$#list-1)] = @tmp[$#list+1..$end];
        @list = @tmp[0..$#list];
    }
    $p = ($p + $len + $skip) % @list;
    $skip++;
}

say "result1: ", $list[0] * $list[1];

File: day10/knot_hash.py
#/usr/bin/env python3

# test data
#myList = [x for x in range(0,5)]
#LENS = (3, 4, 1, 5)

# problem data
myList = [x for x in range(0,256)]
LENS = [147,37,249,1,31,2,226,0,161,71,254,243,183,255,30,70]

p = 0
skip = 0

for myLen in LENS:
    end = p + myLen - 1
    if end < len(myList):
        slice = myList[p:end+1]
        myList[p:end+1] = slice[::-1]
    else:
        tmp = myList + myList
        slice = tmp[p:end+1]
        tmp[p:end+1] = slice[::-1]
        tmp[0:(end-len(myList)+1)] = tmp[len(myList):end+1]
        myList = tmp[0:len(myList)]

    p = (p + myLen + skip) % len(myList)
    skip += 1

print("result1:", myList[0] * myList[1])
